Allow Vue nodes to have their colors changed from selection toolbox

christian-byrne opened the pull request:
## Summary

Fixes https://github.com/Comfy-Org/ComfyUI_frontend/issues/5680 by allowing Vue nodes to properly synchronize color changes with LiteGraph nodes, implementing header darkening and light theme adjustments.

<img width="2496" height="1512" alt="Screenshot from 2025-09-21 20-00-36" src="https://github.com/user-attachments/assets/e3bdf645-1e0b-4d11-9ae5-9401f43e8e96" />

## Changes

- **What**: Implemented color property synchronization between LiteGraph and Vue node rendering systems
- **Core Fix**: Added `nodeData.color` and `nodeData.bgcolor` to [v-memo dependencies](https://vuejs.org/api/built-in-directives.html#v-memo) to trigger re-renders on color changes
- **Color Logic**: Added header darkening using [memoized color adjustments](https://github.com/Comfy-Org/ComfyUI_frontend/blob/main/src/utils/colorUtil.ts) to match LiteGraph's ColorOption system
- **Event System**: Enhanced property change instrumentation in LGraphNode.setColorOption to emit color/bgcolor events

## Review Focus

Vue component reactivity timing - the v-memo fix was critical for immediate color updates. Verify light theme color adjustments match the drawNode monkey patch behavior in app.ts.

## Technical Details

```mermaid
graph TD
    A[User Sets Color] --> B[LGraphNode.setColorOption]
    B --> C[Sets node.color & node.bgcolor]
    C --> D[Triggers property:changed events]
    D --> E[Vue Node Manager Updates]
    E --> F[v-memo Detects Change]
    F --> G[NodeHeader Re-renders]
    G --> H[Header Darkening Applied]

    style A fill:#f9f9f9,stroke:#333,color:#000
    style H fill:#f9f9f9,stroke:#333,color:#000
```

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5720-Allow-Vue-nodes-to-have-their-colors-changed-from-selection-toolbox-2766d73d36508123b441d126a74a54b2) by [Unito](https://www.unito.io)

christian-byrne committed:
fix vue node color change

christian-byrne commented on browser_tests/tests/vueNodes/nodeStates/colors.spec.ts:
```diff
@@ -0,0 +1,52 @@
+    )
+  })
+
+  // TODO: implement loading node colors from workflow in Vue system
+  test.fail('should load node colors from workflow', async ({ comfyPage }) => {
+    await comfyPage.loadWorkflow('nodes/every_node_color')
+    await expect(comfyPage.canvas).toHaveScreenshot(
+      'vue-node-custom-colors-dark-all-colors.png'
+    )
+  })
+
+  // TODO: implement loading node colors from workflow in Vue system
+  test.fail(
```
Expect these to fail until loading custom colors from workflow is implemented.

christian-byrne replied:
They work now

DrJKL commented on src/renderer/extensions/vueNodes/components/NodeHeader.vue:
```diff
@@ -111,6 +114,30 @@ const tooltipConfig = computed(() => {
+  // When header and body colors are the same/similar, darken the header
+  if (nodeData.bgcolor && nodeData.color === nodeData.bgcolor) {
+    // Darken header relative to body (opposite of light theme adjustment)
+    headerColor = adjustColor(nodeData.color, { lightness: -0.15 })
+  }
+
+  // Apply light theme lightening on top of base darkening (same as drawNode monkey patch)
+  if (colorPaletteStore.completedActivePalette.light_theme) {
+    headerColor = adjustColor(headerColor, { lightness: 0.5 })
```
Like this? https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_colors/Relative_colors

DrJKL replied:
If we were able to bind to a CSS variable and derive from it a lighter and a darker shade, we could eliminate all of this logic

christian-byrne committed:
[test] add DOM property assertions for Vue node colors - addresses review feedback

Adds CSS background-color property validation in addition to screenshot comparison
to verify header and body colors are correctly applied to DOM elements.

Co-authored-by: Myestery <Myestery@users.noreply.github.com>

christian-byrne committed:
[refactor] use LGraphNodeProperties for color property tracking - addresses review feedback

Moves color and bgcolor property change tracking to the standard LGraphNodeProperties
system instead of manual event firing in setColorOption. This reduces custom code
and leverages the existing property instrumentation infrastructure.

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
[fix] restore headerStyle binding in NodeHeader template

The :style binding was accidentally removed, causing TypeScript error about
unused headerStyle computed property. This binding is needed for Vue node
header color functionality.

christian-byrne committed:
set w-full on node header

christian-byrne committed:
[feat] fully replicate LiteGraph drawNode color logic in Vue nodes

Now correctly implements all aspects of the LiteGraph drawNode monkey patch:

1. Header colors: Apply opacity + lightness adjustments like LiteGraph
2. Body colors: Apply same adjustments to background as LiteGraph
3. Opacity setting: Support 'Comfy.Node.Opacity' setting from user preferences
4. Light theme: Apply lightness=0.5 to both header and body in light theme

This ensures Vue nodes have pixel-perfect color matching with LiteGraph nodes
across all themes and opacity settings.

christian-byrne committed:
handle opacity setting

Commit:
Update test expectations [skip ci]

christian-byrne committed:
refactor

christian-byrne committed:
change tests from `fail` to normal (should work now)
